Tests/tests/test_Warehouse.cpp

```diff
@@ -0,0 +1,803 @@
+            [&](const uint32_t, const Core::ProxyType<Core::JSON::IElement>& json) {
+                string text;
+                EXPECT_TRUE(json->ToString(text));
+                //Check for error response for the second call - Negative test case
+                EXPECT_EQ(text, string(_T("{"
+                                        "\"jsonrpc\":\"2.0\","
+                                        "\"method\":\"org.rdk.Warehouse.resetDone\","
+                                        "\"params\":{\"success\":false,"
+                                        "\"error\":\"Reset failed\"}"
+                                        "}")));
+                resetDone.SetEvent();
+                return Core::ERROR_NONE;
+            }));
+}
+
+    virtual ~WarehouseResetDeviceFailureTest() override
+    {
+        handler.Unsubscribe(0, _T("resetDone"), _T("org.rdk.Warehouse"), message);
+    }
+};
+
+TEST_F(WarehouseResetDeviceFailureTest, UserFactoryResetDeviceFailure)
+{
+    EXPECT_CALL(iarmBusImplMock, IARM_Bus_Call)
+    .Times(1)
+    .WillOnce(
+        [](const char* ownerName, const char* methodName, void* arg, size_t argLen) {
+            EXPECT_EQ(string(ownerName), string(_T(IARM_BUS_PWRMGR_NAME)));
+            EXPECT_EQ(string(methodName), string(IARM_BUS_PWRMGR_API_UserFactoryReset));
+            //Return error for the Negative test case
+            return IARM_RESULT_INVALID_PARAM;
+        });
+
+    //reset: suppress reboot: true, type: USERFACTORY
+    EXPECT_EQ(Core::ERROR_NONE, handler.Invoke(connection, _T("resetDevice"), _T("{\"suppressReboot\":true,\"resetType\":\"USERFACTORY\"}"), response));
+    EXPECT_EQ(response, _T("{\"success\":true}"));
+    EXPECT_EQ(Core::ERROR_NONE, resetDone.Lock());
+}
+
+TEST_F(WarehouseResetDeviceTest, ColdFactoryResetDevicePwrMgr2RFCEnabled)
+{
+    //Enable Power manager 2
+    plugin->m_isPwrMgr2RFCEnabled = true;
+
+    EXPECT_CALL(wrapsImplMock, system(::testing::_))
+    .Times(4)
+    .WillOnce(::testing::Invoke(
+        [](const char* command) {
+            EXPECT_EQ(string(command), string("sh /lib/rdk/deviceReset.sh coldfactory"));
+            return Core::ERROR_NONE;
+        }))
+    .WillOnce(::testing::Invoke(
+        [](const char* command) {
+            EXPECT_EQ(string(command), string("echo 0 > /opt/.rebootFlag"));
+            return Core::ERROR_NONE;
+        }))
+    .WillOnce(::testing::Invoke(
+        [](const char* command) {
+            EXPECT_EQ(string(command), string("echo `/bin/timestamp` ------------- Rebooting due to Cold Factory Reset process --------------- >> /opt/logs/receiver.log"));
+            return Core::ERROR_NONE;
+        }))
+    .WillOnce(::testing::Invoke(
+        [](const char* command) {
+            EXPECT_EQ(string(command), string("sleep 5; /rebootNow.sh -s PowerMgr_coldFactoryReset -o 'Rebooting the box due to Cold Factory Reset process ...'"));
+            return Core::ERROR_NONE;
+        }));
```

Comment:
"sh /lib/rdk/deviceReset.sh ..." is ok but the rest of commands are too implementation-specific, please replace them with WillRepeatedly to not over-specify the implementation